<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:macro name="so101_ros2_control" params="name use_mock_hardware">

    <ros2_control name="${name}" type="system">
      <xacro:unless value="${use_mock_hardware}">
        <hardware>
            <plugin>so101_control/so101_hardware_interface</plugin>
            <param name="usb_port">/dev/ttyACM0</param>
            <param name="baud_rate">1000000</param>
        </hardware>
      </xacro:unless>
      <xacro:if value="${use_mock_hardware}">
        <hardware>
          <plugin>ros2_control_demo_hardware/GenericSystem</plugin>
        </hardware>
      </xacro:if>

        <joint name="shoulder_pan">
            <param name="id">1</param>
            <param name="type">position</param>
            <param name="position_offset">1.570796</param>
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="torque"/>
            <state_interface name="temperature"/>
            <state_interface name="voltage"/>
        </joint>

        <joint name="shoulder_lift">
            <param name="id">2</param>
            <param name="type">position</param>
            <param name="position_offset">1.570796</param>
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="torque"/>
            <state_interface name="temperature"/>
            <state_interface name="voltage"/>
        </joint>

        <joint name="elbow_flex">
            <param name="id">3</param>
            <param name="type">position</param>
            <param name="position_offset">1.570796</param>
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="torque"/>
            <state_interface name="temperature"/>
            <state_interface name="voltage"/>
        </joint>

        <joint name="wrist_flex">
            <param name="id">4</param>
            <param name="type">position</param>
            <param name="position_offset">1.570796</param>
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="torque"/>
            <state_interface name="temperature"/>
            <state_interface name="voltage"/>
        </joint>

        <joint name="wrist_roll">
            <param name="id">5</param>
            <param name="type">position</param>
            <param name="position_offset">1.570796</param>
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="torque"/>
            <state_interface name="temperature"/>
            <state_interface name="voltage"/>
        </joint>

        <joint name="gripper">
            <param name="id">6</param>
            <param name="type">position</param>
            <param name="position_offset">1.570796</param>
            <command_interface name="position"/>
            <command_interface name="velocity"/>
            <state_interface name="position"/>
            <state_interface name="velocity"/>
            <state_interface name="torque"/>
            <state_interface name="temperature"/>
            <state_interface name="voltage"/>
        </joint>


    </ros2_control> 

    </xacro:macro>


</robot>